weave/trace_server/project_version/project_version.py

```diff
@@ -0,0 +1,136 @@
+import logging
+import threading
+from collections.abc import Callable
+from typing import Any, Optional
+
+import ddtrace
+from cachetools import LRUCache
+
+from weave.trace_server.project_version.clickhouse_project_version import (
+    get_project_data_residence,
+)
+from weave.trace_server.project_version.types import (
+    CallsStorageServerMode,
+    ProjectDataResidence,
+    ReadTable,
+    WriteTarget,
+)
+
+logger = logging.getLogger(__name__)
+
+PER_REPLICA_CACHE_SIZE = 10_000
+
+_resolver: Optional["TableRoutingResolver"] = None
+_resolver_lock = threading.Lock()
+
+
+def init_resolver(
+    ch_client_factory: Callable[[], Any],
+    cache_size: int = PER_REPLICA_CACHE_SIZE,
+) -> None:
+    """Initialize the project version resolver.
+
+    Call this once at application startup. Subsequent calls are ignored
+    (the first initialization wins).
+
+    Args:
+        ch_client_factory: Callable that returns a ClickHouse client.
+        cache_size: Size of the in-memory cache (defaults to 10,000).
+    """
+    global _resolver
+    with _resolver_lock:
```

Comment:
this seems sus. - why not just pass along the ch_client. why need a whole factory and global state

Reply:
its for the in-memory cache